File: 0088-merge-sorted-array/0088-merge-sorted-array.py
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        # start from m+n-1 and put the largest value of num1 and num2 to the end of num1
        # if nums1 has used all its numbers, put nums2 in it
        # if nums2 has used all its numbers, no further action needed
        m -= 1
        n -= 1

        for x in range(len(nums1)-1, -1, -1):
            if m>=0 and n>=0:
                if nums1[m]>nums2[n]:
                    nums1[x] = nums1[m]
                    m -= 1
                else:
                    nums1[x] = nums2[n]
                    n -= 1
            else:
                if m<0:
                    nums1[x] = nums2[n]
                    n -= 1

File: 0088-merge-sorted-array/0088-merge-sorted-array.cpp
class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        // start from the end of 2 arrays
        int i1=m-1, i2=n-1, index=m+n-1;
        while (i1>=0 && i2>=0)
        {
            // put the greater to the last of the array
            if (nums1[i1]>nums2[i2])
            {
                nums1[index] = nums1[i1];
                nums1[i1] = 0;
                --i1;
            }
            else
            {
                nums1[index] = nums2[i2];
                --i2;
            }
            --index;
        }

        // if there are still nums2 remaining
        for (i2; i2>=0; --i2)
        {
            nums1[i2] = nums2[i2];
        }
    }
};